### PR TITLE
Feed misp

### PR DESCRIPTION
**Pull request by maimorag**
## Status
- [x] In Progress

## Related Issues
fixes:[ link to the issue](https://jira-dc.paloaltonetworks.com/browse/XSUP-31708)
https://jira-dc.paloaltonetworks.com/browse/XSUP-31078

## Description
Adding pagination mechanism to ***Feed MISP***

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 1]
```diff
@@ -380,7 +396,7 @@ def fetch_indicators(client: Client,
         create_and_add_relationships(indicator_obj, galaxy_indicators)
 
         indicators.append(indicator_obj)
-
+    demisto.debug(f'{indicators}')
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:399]: Why do we logging all indicators? Lets print more helpful or informative info.

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 2]
```diff
@@ -346,11 +361,12 @@ def fetch_indicators(client: Client,
                      limit: int = -1,
                      is_fetch: bool = True) -> List[Dict]:
     params_dict = clean_user_query(query) if query else build_params_dict(tags, attribute_type)
-
     if limit and limit not in params_dict:
         params_dict['limit'] = limit
-
-    response = client.search_query(params_dict)
+    response = search_query_indicators_pagination(client, params_dict) if is_fetch else client.search_query(params_dict)
+    demisto.debug(f'fetch_indicators response: {response}')
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:367]: We debug the response multiple times, lets consider removing some of the irrelevant debug logs.

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 3]
```diff
@@ -524,10 +540,10 @@ def get_attributes_command(client: Client, args: Dict[str, str], params: Dict[st
     feed_tags = argToList(params.get("feedTags", []))
     query = args.get('query', None)
     attribute_type = argToList(args.get('attribute_type', ''))
-
+    demisto.debug("fetch_indicators starts")
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:543]: Lets add the relevant params we are using to that debug log.

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 4]
```diff
@@ -590,7 +606,7 @@ def main():
             return_results(test_module(client))
         elif command == 'misp-feed-get-indicators':
             return_results(get_attributes_command(client, args, params))
-        elif command == 'fetch-indicators':
+        elif command == 'fetch-indicators':
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:609]: Please remove.

File: Packs/FeedMISP/ReleaseNotes/1_0_30.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+
+#### Integrations
+
+##### MISP Feed
+- Fixed an issue where ***fetch-indicators*** fails if needs to fetch a large amount of indicator, by implanting pagination mechanism
```
Comment (eyalpalo) [inline Packs/FeedMISP/ReleaseNotes/1_0_30.md:5]:
```suggestion
- Fixed an issue where ***fetch-indicators*** timed out when large amount of indicators were fetched.
```

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 6]
```diff
@@ -335,6 +337,19 @@ def update_indicators_iterator(indicators_iterator: List[Dict[str, Any]],
     return []
 
 
+def search_query_indicators_pagination(client: Client, params_dict: Dict[str, Any]) -> Dict[str, Any]:
+    params_dict['page'] = 0
+    response: Dict[str, Dict[str, List]] = {'response': {'Attribute': []}}
+    while True:
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:343]: Lets remove the while True and replace it to a condition relates to the len of _search_query_per_page_ var.

**Review by eyalpalo — CHANGES_REQUESTED** [review]
Nice, lets add some tests also.

**Comment by maimorag** [comment]
@eyalpalo I finished applying the code review notes beside the unit test, and I'm fixing on fixing it

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 7]
```diff
@@ -335,6 +337,20 @@ def update_indicators_iterator(indicators_iterator: List[Dict[str, Any]],
...
+    response: Dict[str, Dict[str, List]] = {'response': {'Attribute': []}}
+    next_page_exist = True
+    while next_page_exist:
+        search_query_per_page = client.search_query(params_dict).get('response', {}).get('Attribute')
+        demisto.debug(f'search_query_per_page:{params_dict["page"]} {search_query_per_page}')
+        if not len(search_query_per_page):
+            next_page_exist = False
+        response['response']['Attribute'].extend(search_query_per_page)
+        params_dict['page'] += 1
+    return response
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:351]:
```suggestion
    search_query_per_page = client.search_query(params_dict).get('response', {}).get('Attribute')
    while len(search_query_per_page):
        demisto.debug(f'search_query_per_page:{params_dict["page"]} {search_query_per_page}')
        response['response']['Attribute'].extend(search_query_per_page)
        params_dict['page'] += 1
        search_query_per_page = client.search_query(params_dict).get('response', {}).get('Attribute')
    return response
```

Pay attention to the indentation as the suggestion made in a text editor.

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP_test.py [hunk 8]
```diff
@@ -448,3 +448,16 @@ def test_update_indicator_fields(
 ])
 def test_get_ip_type(indicator, indicator_type):
     assert get_ip_type(indicator) == indicator_type
+
+
+def test_search_query_indicators_pagination(mocker):
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP_test.py:453]: Please add docstring with Given, When Then as we do for better readability.

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP_test.py [hunk 9]
```diff
@@ -448,3 +448,16 @@ def test_update_indicator_fields(
...
+    client = Client(base_url="example",
+                    authorization="auth",
+                    verify=False,
+                    proxy=False,
+                    timeout=60)
+    mocker.patch.object(Client, '_http_request', return_value={'response': {'Attribute': ['data1', 'data2']}})
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP_test.py:459]: This unit test will have an infinite loop, you return the same response for each request and run in the method until an empty response is received.
Please fix this UT to suit the use case and fix the infinite loop.

**Review by eyalpalo — CHANGES_REQUESTED** [review]

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 10]
```diff
@@ -524,10 +539,9 @@ def get_attributes_command(client: Client, args: Dict[str, str], params: Dict[st
     feed_tags = argToList(params.get("feedTags", []))
     query = args.get('query', None)
     attribute_type = argToList(args.get('attribute_type', ''))
-
+    demisto.debug(f"fetch_indicators starts using the params: {params}")
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:542]:
```suggestion
```

File: Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py [hunk 11]
```diff
@@ -561,9 +575,9 @@ def fetch_attributes_command(client: Client, params: Dict[str, str]) -> List[Dic
     feed_tags = argToList(params.get("feedTags", []))
     attribute_types = argToList(params.get('attribute_types', ''))
     query = params.get('query', None)
-
     indicators = fetch_indicators(client, tags, attribute_types, query, tlp_color,
                                   params.get('url'), reputation, feed_tags)
+    demisto.debug(f"fetch indicator results: {indicators} /n/n fetch_indicators finished")
```
Comment (eyalpalo) [inline Packs/FeedMISP/Integrations/FeedMISP/FeedMISP.py:580]: Lets add the debug log to the fetch indicators itself instead of the main

**Review by eyalpalo — CHANGES_REQUESTED** [review]

**Review by eyalpalo — APPROVED** [review]

**Pull request merged**